server/src/main/java/org/elasticsearch/cluster/routing/allocation/command/MoveAllocationCommand.java

```diff
@@ -122,6 +127,9 @@ public RerouteExplanation execute(RoutingAllocation allocation, boolean explain)
             }
 
             RoutingNode toRoutingNode = allocation.routingNodes().node(toDiscoNode.getId());
+            if (toRoutingNode == null && !toDiscoNode.isDataNode()) {
```

Comment:
can we move this validation up to the start of the method, next to the validation of the fromNode? then all of this is together and we can fail early.